<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="mbkmeans">
<title>An introduction to mbkmeans • mbkmeans</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to mbkmeans">
<meta property="og:description" content="mbkmeans">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mbkmeans</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Category A</h6>
    <a class="dropdown-item" href="../articles/Vignette.html">An introduction to mbkmeans</a>
    <a class="dropdown-item" href="../articles/example_analysis.html">Example data analysis</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/drisso/mbkmeans/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>An introduction to mbkmeans</h1>
                        <h4 data-toc-skip class="author">Yuwei Ni,
Davide Risso, Stephanie Hicks, and Elizabeth Purdom</h4>
            
            <h4 data-toc-skip class="date">Last modified: November 7,
2020; Compiled: November 27, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/drisso/mbkmeans/blob/HEAD/vignettes/Vignette.Rmd" class="external-link"><code>vignettes/Vignette.Rmd</code></a></small>
      <div class="d-none name"><code>Vignette.Rmd</code></div>
    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{mbkmeans vignette}
-->
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the package, please use the following.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"mbkmeans"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette provides an introductory example on how to work with
the <code>mbkmeans</code> package, which contains an implementation of
the mini-batch k-means algorithm proposed in <span class="citation">(Sculley 2010)</span> for large single-cell sequencing
data. This algorithm runs the _k_means iterations on small, random
subsamples of data (``mini batches’’) rather than the entire dataset.
This both speeds up computation and reduces the amount of data that must
be contained in memory.</p>
<p>The main function to be used by the users is <code>mbkmeans</code>.
This is implemented as an S4 generic and methods are implemented for
<code>matrix</code>, <code>Matrix</code>, <code>HDF5Matrix</code>,
<code>DelayedMatrix</code>, <code>SummarizedExperiment</code>, and
<code>SingleCellExperiment</code>.</p>
<p>Most of this work was inspired by the <code>MiniBatchKmeans()</code>
function implemented in the <code>ClusterR</code> R package and we
re-use many of the C++ functions implemented there.</p>
<p>Our main contribution here is to provide an interface to the
<code>DelayedArray</code> and <code>HDF5Array</code> packages so that
<code>mbkmeans</code> algorithm only brings into memory at any time the
“mini-batches” needed for the current iteration, and never requires
loading the entire dataset into memory. This allows the user to run the
mini-batch <em>k</em>-means algorithm on very large datasets that do not
fit entirely in memory. <code>mbkmeans</code> also runs seamlessly on
in-memory matrices provided by the user, and doing so significantly
increases the speed of the algorithm compared to the standard
<em>k</em>-means algorithm (<span class="citation">(Hicks et al.
2020)</span>). A complete comparison of <code>mbkmeans</code> with other
implementations can be found in <span class="citation">(Hicks et al.
2020)</span>.</p>
<p>The motivation for this work is the clustering of large single-cell
RNA-sequencing (scRNA-seq) datasets, and hence the main focus is on
Bioconductor’s <code>SingleCellExperiment</code>and
<code>SummarizedExperiment</code> data container. For this reason,
<code>mbkmeans</code> assumes a data representation typical of genomic
data, in which genes (variables) are in the rows and cells
(observations) are in the column. This is contrary to most other
statistical applications, and notably to the
<code><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">stats::kmeans()</a></code> and
<code><a href="https://rdrr.io/pkg/ClusterR/man/MiniBatchKmeans.html" class="external-link">ClusterR::MiniBatchKmeans()</a></code> functions that assume
observations in rows.</p>
<p>We provide a lower-level <code><a href="../reference/mini_batch.html">mini_batch()</a></code> function that
expects observations in rows and is expected to be a direct replacement
of <code><a href="https://rdrr.io/pkg/ClusterR/man/MiniBatchKmeans.html" class="external-link">ClusterR::MiniBatchKmeans()</a></code> for on-disk data
representations such as <code>HDF5</code> files. The rest of this
document shows the typical use case through the <code><a href="../reference/mbkmeans.html">mbkmeans()</a></code>
interface; users interested in the <code><a href="../reference/mini_batch.html">mini_batch()</a></code> function
should refer to its manual page.</p>
<div class="section level3">
<h3 id="example-dataset">Example dataset<a class="anchor" aria-label="anchor" href="#example-dataset"></a>
</h3>
<p>To illustrate a typical use case, we use the <code>pbmc4k</code>
dataset of the <a href="https://bioconductor.org/packages/release/data/experiment/html/TENxPBMCData.html" class="external-link"><code>TENxPBMCData</code>
package</a>. This dataset contains a set of about 4,000 cells from
peripheral blood from a healthy donor and is expected to contain many
types or clusters of cell.</p>
<p>We would note that <code>mbkmeans</code> is designed for very large
datasets, in particular for datasets large enough that either the full
data matrix can’t be held in memory, or the computations necessary
cannot be computed. This vignette works with a small dataset to allow
the user to quickly follow along and understand the commands. <span class="citation">(Hicks et al. 2020)</span> shows the example of using
<code>mbkmeans</code> with a dataset with 1.3 million cells, which is a
more appropriate size for observing improved memory usage and improved
speed.</p>
<p>First, we load the needed packages.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TENxPBMCData</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">mbkmeans</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PeteHaitch/DelayedMatrixStats" class="external-link">DelayedMatrixStats</a></span><span class="op">)</span></span></code></pre></div>
<p>Note that in this vignette, we do not aim at identifying biologically
meaningful clusters (that would entail a more sophisticated
normalization of data and dimensionality reduction), but instead we only
aim to show how to run mini-batch <em>k</em>-means on a HDF5-backed
matrix.</p>
<p>We normalize the data simply by scaling for the total number of
counts using <code>scater</code> and select the 1,000 most variable
genes and a random set of 100 cells to speed-up computations.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tenx_pbmc4k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxPBMCData/man/TENxPBMCData.html" class="external-link">TENxPBMCData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"pbmc4k"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1034</span><span class="op">)</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">NCOL</a></span><span class="op">(</span><span class="va">tenx_pbmc4k</span><span class="op">)</span><span class="op">)</span>, <span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">tenx_pbmc4k</span><span class="op">[</span>, <span class="va">idx</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#normalization</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowVars.html" class="external-link">rowVars</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">vars</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">vars</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce1000</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">vars</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">]</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">sce1000</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 1000 100 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(1000): ENSG00000090382 ENSG00000204287 ... ENSG00000117748</span></span>
<span><span class="co">##   ENSG00000135968</span></span>
<span><span class="co">## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(12): Sample Barcode ... Date_published sizeFactor</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="mbkmeans">
<code>mbkmeans</code><a class="anchor" aria-label="anchor" href="#mbkmeans"></a>
</h2>
<p>The main function, <code><a href="../reference/mbkmeans.html">mbkmeans()</a></code>, returns a list object
including <code>centroids</code>, <code>WCSS_per_cluster</code> (where
WCSS stands for within-cluster-sum-of-squares),
<code>best_initialization</code>, <code>iters_per_initiazation</code>
and <code>Clusters</code>.</p>
<p>It takes any matrix-like object as input, such as
<code>SummarizedExperiment</code>, <code>SingleCellExperiment</code>,
<code>matrix</code>, <code>DelayedMatrix</code> and
<code>HDF5Matrix</code>.</p>
<p>In this example, the input is a <code>SingleCellExperiment</code>
object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mbkmeans.html">mbkmeans</a></span><span class="op">(</span><span class="va">sce1000</span>, clusters <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                reduceMethod <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                whichAssay <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span></span></code></pre></div>
<p>The number of clusters (such as <em>k</em> in the <em>k</em>-means
algorithm) is set through the <code>clusters</code> argument. In this
case, we set <code>clusters = 5</code> for no particular reason. For
<code>SingleCellExperiment</code> objects, the function provides the
<code>reduceMethod</code> and <code>whichAssay</code> arguments. The
<code>reduceMethod</code> argument should specify the dimensionality
reduction slot to use for the clustering, and the default is “PCA”. Note
that this <em>does not perform</em> PCA but only looks at a slot called
“PCA” already stored in the object. Alternatively, one can specify
<code>whichAssay</code> as the assay to use as input to mini-batch
<em>k</em>-means. This is used only when <code>reduceMethod</code>
option is <code>NA</code>. See <code><a href="../reference/mbkmeans.html">?mbkmeans</a></code> for more
details.</p>
<div class="section level3">
<h3 id="choice-of-arguments">Choice of arguments<a class="anchor" aria-label="anchor" href="#choice-of-arguments"></a>
</h3>
<p>There are additional arguements in <code><a href="../reference/mbkmeans.html">mbkmeans()</a></code> that make
the function more flexible and suitable for more situations.</p>
<div class="section level4">
<h4 id="batch-size">Batch size<a class="anchor" aria-label="anchor" href="#batch-size"></a>
</h4>
<p>The size of the mini batches is set through the
<code>batch_size</code> argument. This argument gives the size of the
samples of data brought into memory at any one time, and thus constrains
the memory usage of the alogrithm.</p>
<p>The <code><a href="../reference/blocksize.html">blocksize()</a></code> function can be used to set the batch
size to the maximum allowed by the available memory. It considers both
the number of columns in the dataset and the amount of RAM on the
current matchine to calculate as large of a batch size as reasonable for
the RAM available to the session. The calculation uses
<code>get_ram()</code> function in <code>benchmarkme</code> package. See
the <code>benchmarkme</code> vignette for more details.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">batchsize</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/blocksize.html">blocksize</a></span><span class="op">(</span><span class="va">sce1000</span><span class="op">)</span></span>
<span><span class="va">batchsize</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 100</span></span></code></pre>
<p>In this simple example, because the whole data fits in memory, the
default batch size would be a single batch of size 100.</p>
<p><span class="citation">(Hicks et al. 2020)</span> provides a
comparison of the effect of batch size for large datasets (i.e. where
the full data cannot be stored in memory). Their results show the
algorithm to be robust to a large range of batch sizes (1,000-10,000
cells): accuracy of the results were equivalent if the batch size was
above 500-1000 cells, with no noticable advantage in memory usage or
speed as long as the batches were smaller than roughly 10,000 cells
(corresponding to a subsample of the full data matrix whose memory foot
print is not noticeable on most modern computers). For this reason, we
would err on the larger range of batch size suggested in <span class="citation">(Hicks et al. 2020)</span> – 10,000 cells. Furthermore,
since the initialization uses by default the same number of cells as
batch size, this gives a robust sample for determining the initial start
point.</p>
<p>We would note that it is better to make choices on batch size based
on choosing the <em>absolute</em> number of cells in a batch rather than
choosing batch size based on a percentage of the overall sample size. If
batch size is chosen based on the percentage of cells, very large
datasets will use very large batches of cells, negating the memory and
speed gains. Moreover such large batches are not needed for good
performance of the algorithm. The continual resampling of the batches
works like stochastic gradient descent and allows for (local)
minimization of the objective function <span class="citation">(Sculley
2010)</span> with relatively small batch sizes, as demonstrated on
single-cell data in the work of <span class="citation">(Hicks et al.
2020)</span>. If anything, the size of the batch should be governed more
by the complexity of the problem (e.g. the number of underlying true
clusters or subtypes), rather than the number of overall cells in the
dataset.</p>
</div>
<div class="section level4">
<h4 id="initialization">Initialization<a class="anchor" aria-label="anchor" href="#initialization"></a>
</h4>
<p>The performance of mini-batch <em>k</em>-means greatly depends on the
process of initialization. We implemented two different initialization
methods:</p>
<ol style="list-style-type: decimal">
<li>Random initialization, as in regular <em>k</em>-means;</li>
<li>
<code>kmeans++</code>, as proposed in <span class="citation">(Arthur
and Vassilvitskii 2007)</span>. The default is “kmeans++”.</li>
</ol>
<p>The percentage of data to use for the initialization centroids is set
through the <code>init_fraction</code> argument, which should be larger
than 0 and less than 1. The default value is given so that the
proportion matches the value of <code>batch_size</code>, converted to a
proportion of the data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_random</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mbkmeans.html">mbkmeans</a></span><span class="op">(</span><span class="va">sce1000</span>, clusters <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                reduceMethod <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                whichAssay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>                initializer <span class="op">=</span> <span class="st">"random"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">Clusters</span>, <span class="va">res_random</span><span class="op">$</span><span class="va">Clusters</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    </span></span>
<span><span class="co">##      1  2  3  4  5</span></span>
<span><span class="co">##   1  0  0  0  0  1</span></span>
<span><span class="co">##   2  0  0  1  0  0</span></span>
<span><span class="co">##   3  4  0 23  0  0</span></span>
<span><span class="co">##   4  0  0  0  0  4</span></span>
<span><span class="co">##   5  0  5  0 14 48</span></span></code></pre>
<p>Note that large values of <code>init_fraction</code> will result in
that proportion of the data being held in memory and used in the
initialization of the <code>kmeans++</code> algorithm. Therefore, large
values of <code>init_fraction</code> will require a great deal of memory
– potentially more than that used by the actual clustering algorithm.
Indeed keeping the fraction used constant for larger and larger datasets
will result in commiserate numbers of cells being used in the
initialization and increasing the memory usage (similar to the issues
discussed above for batch size). Therefore, we recommend that users keep
the default and make changes to the <code>batch_size</code> parameter as
desired, ensuring consistent memory usage across both the initialization
stage and the actual clustering stage of the algorithm. As mentioned in
the discussion above on <code>batch_size</code>, for this reason we
would recommend <code>batch_size</code> on the larger range suggested as
reasonable in <span class="citation">(Hicks et al. 2020)</span> so as to
improve the initialization.</p>
</div>
</div>
<div class="section level3">
<h3 id="running-mbkmeans-with-multiple-values-of-k">Running <code>mbkmeans</code> with multiple values of <span class="math inline">\(k\)</span><a class="anchor" aria-label="anchor" href="#running-mbkmeans-with-multiple-values-of-k"></a>
</h3>
<p>The main parameter to set in <span class="math inline">\(k\)</span>-means and its variants is the number of
clusters <span class="math inline">\(k\)</span>. <code>mbkmeans</code>
is quick enough to rerun the clustering algorithm with different number
of clusters even in very large datasets.</p>
<p>Here, we apply <code>mbkmeans</code> with <span class="math inline">\(k\)</span> from 5 to 15 and select the number of
clusters using the elbow method (i.e., the value that corresponds to the
point of inflection on the curve). We note that this is just a rule of
thumb for selecting the number of clusters, and many different methods
exist to decide the appropriate value of <span class="math inline">\(k\)</span>.</p>
<p>To speed up the computations, we will cluster on the top 20 PCs.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce1000</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce1000</span>, ncomponents<span class="op">=</span><span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">15</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ks</span>, <span class="kw">function</span><span class="op">(</span><span class="va">k</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/mbkmeans.html">mbkmeans</a></span><span class="op">(</span><span class="va">sce1000</span>, clusters <span class="op">=</span> <span class="va">k</span>,</span>
<span>             reduceMethod <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>             calc_wcss <span class="op">=</span> <span class="cn">TRUE</span>, num_init<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">wcss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">res</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">WCSS_per_cluster</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ks</span>, <span class="va">wcss</span>, type <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Vignette_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>From the plot, it seems that <code>k = 12</code> could be a
reasonable value.</p>
</div>
</div>
<div class="section level2">
<h2 id="comparison-with-k-means">Comparison with <em>k</em>-means<a class="anchor" aria-label="anchor" href="#comparison-with-k-means"></a>
</h2>
<p>Note that if we set <code>init_fraction = 1</code>,
<code>initializer = "random"</code>, and
<code>batch_size = ncol(x)</code>, we recover the classic
<em>k</em>-means algorithm.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_full</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mbkmeans.html">mbkmeans</a></span><span class="op">(</span><span class="va">sce1000</span>, clusters <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                     reduceMethod <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                     whichAssay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>                     initializer <span class="op">=</span> <span class="st">"random"</span>,</span>
<span>                     batch_size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce1000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">res_classic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sce1000</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                      centers <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">res_full</span><span class="op">$</span><span class="va">Clusters</span>, <span class="va">res_classic</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    </span></span>
<span><span class="co">##      1  2  3  4  5</span></span>
<span><span class="co">##   1  0  7  0  7  0</span></span>
<span><span class="co">##   2  0  0 14 43  0</span></span>
<span><span class="co">##   3  2  0  0  0  0</span></span>
<span><span class="co">##   4 16  0  0  0 10</span></span>
<span><span class="co">##   5  0  0  0  1  0</span></span></code></pre>
<p>Note however, that since the two algorithms start from different
random initializations, they will not always converge to the same
solution. Furthremore, if the algorithm uses batches smaller than the
full dataset (<code>batch_size&lt;ncol(x)</code>), then
<code>mbkmeans</code> will not converge to the same result as
<code>kmeans</code> even with the same initialization. This is due to
the fact that <code>kmeans</code> and <code>mbkmeans</code> have
different search paths for an optimum and the problem is non-convex.</p>
<p>For a comparison of memory usage, speed, and accuracy of
<code>mbkmeans</code> with <code>kmeans</code> for the non-trivial case
where the algorithm uses batches, see <span class="citation">(Hicks et
al. 2020)</span>, where it is demonstrated that <code>mbkmeans</code>
provides significant speed improvements and uses less memory than
<code>kmeans</code>, without loss of accuracy.</p>
</div>
<div class="section level2">
<h2 id="use-with-bluster">Use with bluster<a class="anchor" aria-label="anchor" href="#use-with-bluster"></a>
</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.15/bluster" class="external-link">bluster</a></em>
package provides a flexible and extensible framework for clustering in
Bioconductor packages/workflows. The <code><a href="../reference/clusterRows.html">clusterRows()</a></code> function
controls dispatch to different clustering algorithms. In this case, we
will use the mini-batch <em>k</em>-means algorithm to cluster cells into
cell populations based on their principal component analysis (PCA)
coordinates.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce1000</span>, <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 100  20</span></span></code></pre>
<p>In the following three scenarios, we pass the number of cluster
centers as a number using the <code>centers=5</code> argument. In
addition, we can also pass other <code>mbkmeans</code> arguments
(e.g. <code>batch_size=10</code>). In the first two cases, we only
return the cluster labels. However, in the third scenario, we ask for
the full <code>mbkmeans</code> output.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/clusterRows.html">clusterRows</a></span><span class="op">(</span><span class="va">mat</span>, <span class="fu"><a href="../reference/MbkmeansParam.html">MbkmeansParam</a></span><span class="op">(</span>centers<span class="op">=</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   [1] 2 5 5 3 5 4 3 1 1 3 5 2 5 1 1 1 3 5 1 3 2 1 5 5 1 3 1 1 5 1 3 5 4 1 2 1 4</span></span>
<span><span class="co">##  [38] 5 5 4 5 4 5 4 5 3 1 3 5 5 5 2 1 3 2 5 4 1 4 3 5 4 1 1 3 5 2 4 5 4 1 5 4 1</span></span>
<span><span class="co">##  [75] 5 3 1 5 3 1 2 1 4 1 2 5 4 5 1 1 5 1 4 2 4 5 2 5 1 2</span></span>
<span><span class="co">## Levels: 1 2 3 4 5</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clusterRows.html">clusterRows</a></span><span class="op">(</span><span class="va">mat</span>, <span class="fu"><a href="../reference/MbkmeansParam.html">MbkmeansParam</a></span><span class="op">(</span>centers<span class="op">=</span><span class="fl">5</span>, batch_size<span class="op">=</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   [1] 3 3 3 3 3 3 3 5 5 3 3 1 3 5 5 5 3 3 5 3 4 5 3 3 5 3 5 5 3 5 3 3 4 5 1 5 4</span></span>
<span><span class="co">##  [38] 3 3 3 3 3 3 3 3 3 5 3 3 3 3 1 5 3 1 3 3 5 4 3 3 4 5 5 3 3 3 3 3 3 2 3 3 5</span></span>
<span><span class="co">##  [75] 3 3 5 3 3 5 4 5 3 5 1 3 4 3 5 5 3 5 3 3 3 3 3 3 5 4</span></span>
<span><span class="co">## Levels: 1 2 3 4 5</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clusterRows.html">clusterRows</a></span><span class="op">(</span><span class="va">mat</span>, <span class="fu"><a href="../reference/MbkmeansParam.html">MbkmeansParam</a></span><span class="op">(</span>centers<span class="op">=</span><span class="fl">5</span>, batch_size<span class="op">=</span><span class="fl">10</span><span class="op">)</span>, full <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $clusters</span></span>
<span><span class="co">##   [1] 5 5 5 5 5 5 5 1 1 5 5 2 5 1 1 1 5 5 3 5 5 1 5 5 1 5 1 1 5 1 5 5 5 1 2 1 5</span></span>
<span><span class="co">##  [38] 5 5 5 5 5 5 5 5 5 1 5 5 5 5 2 1 5 2 5 5 1 5 5 5 5 1 1 5 5 5 5 5 5 1 5 5 1</span></span>
<span><span class="co">##  [75] 5 5 1 5 5 1 5 1 5 1 2 5 5 5 1 1 5 1 5 5 5 5 4 5 1 5</span></span>
<span><span class="co">## Levels: 1 2 3 4 5</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $objects</span></span>
<span><span class="co">## $objects$mbkmeans</span></span>
<span><span class="co">## $objects$mbkmeans$centroids</span></span>
<span><span class="co">##            [,1]       [,2]      [,3]       [,4]      [,5]       [,6]       [,7]</span></span>
<span><span class="co">## [1,] -15.010417 -0.7962079 -4.968180  2.7679840 -0.605855 -0.2194566  1.4958129</span></span>
<span><span class="co">## [2,]   1.398788 15.2268393  4.914318  8.7031508 -0.624902 -2.6562225  0.8620272</span></span>
<span><span class="co">## [3,] -17.543300 -0.3406707 -6.138011  4.3145411  1.079512  0.6031462 -3.6588221</span></span>
<span><span class="co">## [4,]   3.505305  9.0542211  4.008002  0.8464335 -1.494348 -5.7764435  3.4572454</span></span>
<span><span class="co">## [5,]   4.953126 -5.2662492  1.885661 -0.1000841  1.303848 -0.6517419  0.3327930</span></span>
<span><span class="co">##            [,8]       [,9]      [,10]      [,11]     [,12]      [,13]</span></span>
<span><span class="co">## [1,] -0.6686381  0.7806568 -1.1359341  1.1528491 -1.407013  1.5457278</span></span>
<span><span class="co">## [2,] -3.2219319 -3.0456175 -0.2216553 -1.8928053  5.403923 -2.9907297</span></span>
<span><span class="co">## [3,]  8.2619290 -4.0679099 -3.2909212  5.7171778  8.625866  4.8241775</span></span>
<span><span class="co">## [4,] -3.3455593 -2.0061982 -1.3108509 -9.4799142  1.971986  9.3333830</span></span>
<span><span class="co">## [5,] -0.4513868 -1.7769361 -1.2359026  0.4449523  0.151892 -0.9781383</span></span>
<span><span class="co">##           [,14]      [,15]      [,16]       [,17]      [,18]     [,19]</span></span>
<span><span class="co">## [1,]  1.6981873  3.2334236  0.7828917 -0.11174810 -0.7921813 -2.095090</span></span>
<span><span class="co">## [2,] -1.0326219 -0.3752647 -2.2624371 -2.48013343 -2.6617969  2.687306</span></span>
<span><span class="co">## [3,]  0.2712364 -1.0289170  5.9773702  2.34843263 -1.5557200  3.698316</span></span>
<span><span class="co">## [4,]  0.8668148  1.5455810  1.8839526  4.15320464 -2.9828210 -2.413375</span></span>
<span><span class="co">## [5,]  2.0870510 -1.4041611  0.3987755  0.06157393 -0.6216171  0.435778</span></span>
<span><span class="co">##            [,20]</span></span>
<span><span class="co">## [1,]  0.12527754</span></span>
<span><span class="co">## [2,]  2.21443762</span></span>
<span><span class="co">## [3,] -2.67774700</span></span>
<span><span class="co">## [4,] -0.06607878</span></span>
<span><span class="co">## [5,]  0.77234827</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $objects$mbkmeans$WCSS_per_cluster</span></span>
<span><span class="co">## numeric(0)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $objects$mbkmeans$best_initialization</span></span>
<span><span class="co">## [1] 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $objects$mbkmeans$iters_per_initialization</span></span>
<span><span class="co">##      [,1]</span></span>
<span><span class="co">## [1,]   17</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $objects$mbkmeans$Clusters</span></span>
<span><span class="co">##   [1] 5 5 5 5 5 5 5 1 1 5 5 2 5 1 1 1 5 5 3 5 5 1 5 5 1 5 1 1 5 1 5 5 5 1 2 1 5</span></span>
<span><span class="co">##  [38] 5 5 5 5 5 5 5 5 5 1 5 5 5 5 2 1 5 2 5 5 1 5 5 5 5 1 1 5 5 5 5 5 5 1 5 5 1</span></span>
<span><span class="co">##  [75] 5 5 1 5 5 1 5 1 5 1 2 5 5 5 1 1 5 1 5 5 5 5 4 5 1 5</span></span></code></pre>
<p>While many of the times clustering will be performed on in-memory
data (for example after PCA), the <code>clusterRows</code> function also
accepts any matrix-like object (in-memory or on-disk data
representation) that <code>mbkmeans</code> accepts.</p>
<p>For example, we can cluster using <code>mbkmeans</code> on the
<code>logcounts</code> assay in the <code>sce1000</code> object (similar
to above), which is a <code>DelayedMatrix</code> object from the <em><a href="https://bioconductor.org/packages/3.15/DelayedArray" class="external-link">DelayedArray</a></em>
package.</p>
<p><strong>Note</strong>: We transpose the matrix as the
<code>clusterRows</code> function expects observations along the rows
and variables along the columns.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sce1000</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## &lt;1000 x 100&gt; sparse matrix of class DelayedMatrix and type "double":</span></span>
<span><span class="co">##                      [,1]      [,2]      [,3] ...     [,99]    [,100]</span></span>
<span><span class="co">## ENSG00000090382  0.000000  1.482992  0.000000   . 5.4330604 0.0000000</span></span>
<span><span class="co">## ENSG00000204287  2.198815  0.000000  0.000000   . 5.3387971 0.0000000</span></span>
<span><span class="co">## ENSG00000163220  1.762993  1.482992  0.000000   . 2.1176311 2.0642508</span></span>
<span><span class="co">## ENSG00000143546  0.000000  1.135438  1.113099   . 0.3825221 0.0000000</span></span>
<span><span class="co">## ENSG00000019582  3.402730  1.762861  1.733874   . 6.2009459 2.0642508</span></span>
<span><span class="co">##             ...         .         .         .   .         .         .</span></span>
<span><span class="co">## ENSG00000145779 0.0000000 0.6766523 0.0000000   . 0.9342195 0.0000000</span></span>
<span><span class="co">## ENSG00000106948 0.0000000 1.1354382 0.0000000   . 0.9342195 1.3735556</span></span>
<span><span class="co">## ENSG00000155660 0.0000000 0.6766523 0.0000000   . 0.3825221 1.3735556</span></span>
<span><span class="co">## ENSG00000117748 0.0000000 0.0000000 0.0000000   . 0.0000000 0.0000000</span></span>
<span><span class="co">## ENSG00000135968 0.0000000 0.0000000 0.0000000   . 0.3825221 0.0000000</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clusterRows.html">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sce1000</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="../reference/MbkmeansParam.html">MbkmeansParam</a></span><span class="op">(</span>centers<span class="op">=</span><span class="fl">4</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   [1] 2 2 2 1 2 2 1 3 3 1 2 2 2 3 3 3 1 2 3 1 2 3 2 2 3 1 3 3 2 3 1 2 2 3 2 3 2</span></span>
<span><span class="co">##  [38] 2 2 2 2 2 2 2 2 1 3 1 2 2 2 2 3 1 2 4 2 3 2 1 2 2 3 3 1 2 2 2 2 2 3 2 2 3</span></span>
<span><span class="co">##  [75] 2 1 3 2 1 3 2 3 2 3 2 2 2 2 3 3 2 3 2 2 2 2 2 2 3 2</span></span>
<span><span class="co">## Levels: 1 2 3 4</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="working-with-on-disk-data-in-bioconductor">Working with on-disk data in Bioconductor<a class="anchor" aria-label="anchor" href="#working-with-on-disk-data-in-bioconductor"></a>
</h2>
<p>This vignette is focused on the use of the <code>mbkmeans</code>
package. For the users interested in learning more about working with
<code>HDF5</code> files and on-disk data in Bioconductor, we recommend
the excellent <a href="https://petehaitch.github.io/BioC2020_DelayedArray_workshop/articles/Effectively_using_the_DelayedArray_framework_for_users.html" class="external-link">DelayedArray
workshop</a>, presented by Peter Hickey at <a href="https://bioc2020.bioconductor.org/" class="external-link">Bioc2020</a>.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-arthur2007k" class="csl-entry">
Arthur, David, and Sergei Vassilvitskii. 2007. <span>“K-Means++: The
Advantages of Careful Seeding.”</span> In <em>Proceedings of the
Eighteenth Annual ACM-SIAM Symposium on Discrete Algorithms</em>,
1027–35. Society for Industrial; Applied Mathematics.
</div>
<div id="ref-Hicks2020" class="csl-entry">
Hicks, Stephanie C., Ruoxi Liu, Yuwei Ni, Elizabeth Purdom, and Davide
Risso. 2020. <span>“Mbkmeans: Fast Clustering for Single Cell Data Using
Mini-Batch k-Means.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2020.05.27.119438" class="external-link">https://doi.org/10.1101/2020.05.27.119438</a>.
</div>
<div id="ref-sculley2010web" class="csl-entry">
Sculley, David. 2010. <span>“Web-Scale k-Means Clustering.”</span> In
<em>Proceedings of the 19th International Conference on World Wide
Web</em>, 1177–78. ACM.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yuwei Ni, Davide Risso, Stephanie Hicks, Elizabeth Purdom.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
