<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This is an implementation of the mini-batch k-means algorithm of
  Sculley (2010) for large single cell sequencing data with the
  dimensionality reduction results as input in the reducedDim() slot."><title>Mini-Batch k-means for large single cell sequencing data — mbkmeans • mbkmeans</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Mini-Batch k-means for large single cell sequencing data — mbkmeans"><meta property="og:description" content="This is an implementation of the mini-batch k-means algorithm of
  Sculley (2010) for large single cell sequencing data with the
  dimensionality reduction results as input in the reducedDim() slot."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mbkmeans</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Category A</h6>
    <a class="dropdown-item" href="../articles/Vignette.html">An introduction to mbkmeans</a>
    <a class="dropdown-item" href="../articles/example_analysis.html">Example data analysis</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/drisso/mbkmeans/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Mini-Batch k-means for large single cell sequencing data</h1>
      <small class="dont-index">Source: <a href="https://github.com/drisso/mbkmeans/blob/HEAD/R/AllGenerics.R" class="external-link"><code>R/AllGenerics.R</code></a>, <a href="https://github.com/drisso/mbkmeans/blob/HEAD/R/kmeans.R" class="external-link"><code>R/kmeans.R</code></a></small>
      <div class="d-none name"><code>mbkmeans.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This is an implementation of the mini-batch k-means algorithm of
  Sculley (2010) for large single cell sequencing data with the
  dimensionality reduction results as input in the reducedDim() slot.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mbkmeans</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SummarizedExperiment</span></span>
<span><span class="fu">mbkmeans</span><span class="op">(</span><span class="va">x</span>, whichAssay <span class="op">=</span> <span class="fl">1</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for SingleCellExperiment</span></span>
<span><span class="fu">mbkmeans</span><span class="op">(</span><span class="va">x</span>, reduceMethod <span class="op">=</span> <span class="st">"PCA"</span>, whichAssay <span class="op">=</span> <span class="fl">1</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for LinearEmbeddingMatrix</span></span>
<span><span class="fu">mbkmeans</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ANY</span></span>
<span><span class="fu">mbkmeans</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">clusters</span>,</span>
<span>  batch_size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">NCOL</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  max_iters <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  num_init <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  init_fraction <span class="op">=</span> <span class="va">batch_size</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">NCOL</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>,</span>
<span>  initializer <span class="op">=</span> <span class="st">"kmeans++"</span>,</span>
<span>  compute_labels <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  calc_wcss <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  early_stop_iter <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  CENTROIDS <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">1e-04</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>The object on which to run mini-batch k-means. It can be a
matrix-like object (e.g., matrix, Matrix, DelayedMatrix, HDF5Matrix) with
genes in the rows and samples in the columns. Specialized methods are
defined for SummarizedExperiment and SingleCellExperiment.</p></dd>


<dt>...</dt>
<dd><p>passed to `blockApply`.</p></dd>


<dt>whichAssay</dt>
<dd><p>The assay to use as input to mini-batch k-means. If x is a
SingleCellExperiment, this is ignored unless <code>reduceMethod = NA</code>.</p></dd>


<dt>reduceMethod</dt>
<dd><p>Name of dimensionality reduction results to use as input
to mini-batch k-means. Set to NA to use the full matrix.</p></dd>


<dt>clusters</dt>
<dd><p>the number of clusters</p></dd>


<dt>batch_size</dt>
<dd><p>the size of the mini batches. By default, it equals the
minimum between the number of observations and 500.</p></dd>


<dt>max_iters</dt>
<dd><p>the maximum number of clustering iterations</p></dd>


<dt>num_init</dt>
<dd><p>number of times the algorithm will be run with different
centroid seeds</p></dd>


<dt>init_fraction</dt>
<dd><p>proportion of data to use for the initialization
centroids (applies if initializer is <em>kmeans++</em> ). Should be a float
number between 0.0 and 1.0. By default, it uses the relative batch size.</p></dd>


<dt>initializer</dt>
<dd><p>the method of initialization. One of <em>kmeans++</em> and
<em>random</em>. See details for more information</p></dd>


<dt>compute_labels</dt>
<dd><p>logcical indicating whether to compute the final cluster
labels.</p></dd>


<dt>calc_wcss</dt>
<dd><p>logical indicating whether the per-cluster WCSS
is computed. Ignored if `compute_labels = FALSE`.</p></dd>


<dt>early_stop_iter</dt>
<dd><p>continue that many iterations after calculation of the
best within-cluster-sum-of-squared-error</p></dd>


<dt>verbose</dt>
<dd><p>either TRUE or FALSE, indicating whether progress is printed
during clustering</p></dd>


<dt>CENTROIDS</dt>
<dd><p>a matrix of initial cluster centroids. The rows of the
CENTROIDS matrix should be equal to the number of clusters and the columns
should be equal to the columns of the data</p></dd>


<dt>tol</dt>
<dd><p>a float number. If, in case of an iteration (iteration &gt; 1 and
iteration &lt; max_iters) 'tol' is greater than the squared norm of the
centroids, then kmeans has converged</p></dd>


<dt>BPPARAM</dt>
<dd><p>See the `BiocParallel` package. Only the label assignment is
done in parallel.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list with the following attributes: centroids, WCSS_per_cluster,
  best_initialization, iters_per_initialization.</p>


<p>a list with the following attributes: centroids, WCSS_per_cluster,
 best_initialization, iters_per_initialization</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The implementation is largely based on the
  <code><a href="https://rdrr.io/pkg/ClusterR/man/MiniBatchKmeans.html" class="external-link">MiniBatchKmeans</a></code> function of the <code>ClusterR</code>
  package. The contribution of this package is to provide support for on-disk
  data representations such as HDF5, through the use of <code>DelayedMatrix</code>
  and <code>HDF5Matrix</code> objects, as well as for sparse data representation
  through the classes of the <code>Matrix</code> package. We also provide
  high-level methods for objects of class <code>SummarizedExperiment</code>,
  <code>SingleCellExperiment</code>, and <code>LinearEmbeddingMatrix</code>.</p>
<p>This function performs k-means clustering using mini batches.</p>
<p><strong>kmeans++</strong>: kmeans++ initialization. Reference :
http://theory.stanford.edu/~sergei/papers/kMeansPP-soda.pdf AND
http://stackoverflow.com/questions/5466323/how-exactly-does-k-means-work</p>
<p><strong>random</strong>: random selection of data rows as initial centroids</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Sculley. Web-Scale K-Means Clustering. WWW 2010, April 26–30,
  2010, Raleigh, North Carolina, USA. ACM 978-1-60558-799-8/10/04.</p>
<p>https://github.com/mlampros/ClusterR</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Lampros Mouselimis and Yuwei Ni</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: MatrixGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: matrixStats</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MatrixGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedSds, rowWeightedVars</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomicRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: stats4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘BiocGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Filter, Find, Map, Position, Reduce, anyDuplicated, append,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.data.frame, basename, cbind, colnames, dirname, do.call,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     tapply, union, unique, unsplit, which.max, which.min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: S4Vectors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘S4Vectors’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     I, expand.grid, unname</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biobase</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to Bioconductor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘Biobase’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:MatrixGenerics’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     anyMissing, rowMedians</span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">mbkmeans</span><span class="op">(</span><span class="va">se</span>, clusters <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $centroids</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             [,1]       [,2]       [,3]       [,4]      [,5]       [,6]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] -0.86233366 -0.1930144  0.6247238 -0.1342908 0.0651359 -0.5407030</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] -0.09849951 -0.9753502 -0.2598674  0.1835491 0.9668127 -0.6430539</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            [,7]        [,8]       [,9]      [,10]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] -0.4014151  0.07637861 -0.7058427 -0.7009525</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,]  1.2406141 -1.71585792  0.6492191  1.1677006</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $WCSS_per_cluster</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> numeric(0)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $best_initialization</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $iters_per_initialization</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [,1]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Clusters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 1 1 2 1 1 1 1 2 1 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">mbkmeans</span><span class="op">(</span><span class="va">sce</span>, clusters <span class="op">=</span> <span class="fl">2</span>, reduceMethod <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $centroids</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [,1]       [,2]       [,3]       [,4]       [,5]        [,6]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 0.6842993  0.1386736 -0.4185955 0.44005471 -0.3672732 -0.06780258</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 1.0046457 -0.6963494 -0.1473858 0.05162198  0.4315053  0.67613734</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [,7]       [,8]       [,9]      [,10]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 0.3516456  0.1282004 -0.4356472  0.1168733</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 1.4198397 -1.4370986 -0.2415942 -0.9297087</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $WCSS_per_cluster</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> numeric(0)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $best_initialization</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $iters_per_initialization</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [,1]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Clusters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 1 1 1 2 1 1 1 1 1 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">x</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">mbkmeans</span><span class="op">(</span><span class="va">x</span>,clusters <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $centroids</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             [,1]       [,2]       [,3]       [,4]       [,5]       [,6]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] -0.81011435 -0.8828415  1.0545466 -0.7624309 -0.1317222  0.4530511</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] -0.07964226  0.1615282 -0.9620170  0.1830597 -0.6721411 -1.1632454</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,]  0.42792790  0.6814623  0.2438877 -0.6567995  0.7017171 -0.5291350</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             [,7]       [,8]       [,9]       [,10]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]  1.13921503 -0.2021040 -0.2100428 -0.38370619</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] -0.08921928  0.3774196  0.6599757 -0.53591208</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,] -0.78414344 -0.3207443  0.1097594  0.01639013</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $WCSS_per_cluster</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> numeric(0)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $best_initialization</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $iters_per_initialization</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [,1]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Clusters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 1 3 3 3 2 3 3 2 1 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Yuwei Ni, Davide Risso, Stephanie Hicks, Elizabeth Purdom.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

